dds/DCPS/XTypes/TypeObject.h

```diff
@@ -1949,6 +1963,19 @@ bool operator<<(Serializer& strm, const XTypes::Sequence<T>& seq)
 template<typename T>
 bool operator>>(Serializer& strm, XTypes::Sequence<T>& seq)
 {
+  size_t total_size = 0;
+  if (!strm.read_delimiter(total_size)) {
+    return false;
+  }
+  // special case for compatibility with older versions that encoded this
+  // sequence incorrectly - if the DHeader was read as a 0, it's an empty
+  // sequence although it should have been encoded as DHeader (4) + Length (0)
+  if (total_size == 0) {
+    seq.length(0);
+    return true;
+  }
+
+  const size_t end_of_seq = strm.pos() + total_size;
```

Comment:
A couple of questions I have with this function.
(1) When user provides `-DCPSTypeObjectEncoding ReadOldFormat` option, does that mean this function should skip reading DHEADER, because from my understanding, the old format is the incorrect one that doesn't have DHEADER? If so, this function still reads DHEADER regardless of the encoding type given to `-DCPSTypeObjectEncoding`.

(2) In case the received data doesn't have DHEADER and the sequence has a length greater than 0, it seems to me it's possible that the function could incorrectly construct `seq` and still return true. This's because `total_size` is actually the length of the sequence, and `length` is read from the first 4 bytes of the first element of the sequence, and so on. So seems like the function could read wrong data into `seq` but still returns true?

Reply:
(1) `ReadOldFormat` means that we need to be compatible with both peers that are correct and peers that use the wrong encoding.  In this mode (and actually in all modes since there's no drawback) it will allow a 0 in this position in the stream.

(2) Fortunately for us, this case does not happen.  The version that has this bug always sends the `sequence<TypeIdentfierWithSize> dependent_typeids;` with length 0, and that's the only sequence-of-struct in `TypeInformation`.